### PR TITLE
RD-5753 Fully migrating `app/utils` directory to typescript

### PR DESCRIPTION
**Pull request by Vorbert-Kruk**
## Description

Changes consists of:
- Migrating `app/utils` directory to typescript
- Migrating away from implicit lodash imports
- Applying [suggestion from #2289 ](https://github.com/cloudify-cosmo/cloudify-stage/pull/2289#discussion_r1035754332)
- Simplifying properties assignment in class constructor, by using `private` keyword 🎉 

## Screenshots / Videos

N/A

## Checklist

- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I followed the [UI Testing Guidelines](https://cloudifysource.atlassian.net/l/cp/udCVfvrx) and verified that all tests/checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests

 

N/A

## Documentation

N/A

### CODE REVIEW

File: app/utils/WidgetParamsHandler.ts [hunk 1]
```diff
@@ -58,28 +61,31 @@ export default class WidgetParamsHandler {
...
-            } catch (e) {
-                log.error('Error processing fetch params', e);
-                throw new Error('Error processing fetch params', e);
+                this.fetchParams.filterParams = this.widget.definition.fetchParams(
+                    this.widget as Widget<Record<string, keyof Stage.Types.GridParams>>,
+                    this.toolbox
+                );
+            } catch (err) {
+                log.error('Error processing fetch params', err);
+                throw new Error('Error processing fetch params');
```
Comment (Vorbert-Kruk) [inline app/utils/WidgetParamsHandler.ts:80]: ### Note
I've removed `err` argument as typescript was instructing me that `new Error(...)` can be provided with only 1 argument

File: app/utils/WidgetParamsHandler.ts [hunk 2]
```diff
@@ -88,19 +94,19 @@ export default class WidgetParamsHandler {
...
-        if (_.isFunction(this.widget.definition.mapGridParams)) {
+        if (isFunction(this.widget.definition.mapGridParams)) {
             try {
                 params = this.widget.definition.mapGridParams(this.fetchParams.gridParams);
-            } catch (e) {
-                log.error('Error processing match grid params', e);
-                throw new Error('Error processing match grid params', e);
+            } catch (err) {
+                log.error('Error processing match grid params', err);
+                throw new Error('Error processing match grid params');
```
Comment (Vorbert-Kruk) [inline app/utils/WidgetParamsHandler.ts:109]: ### Note
I've removed `err` argument as typescript was instructing me that `new Error(...)` can be provided with only 1 argument

File: app/utils/WidgetParamsHandler.ts [hunk 3]
```diff
@@ -58,28 +61,31 @@ export default class WidgetParamsHandler {
...
     runFetchParamsIfNeeded() {
-        if (_.isFunction(this.widget.definition.fetchParams)) {
+        if (isFunction(this.widget.definition.fetchParams)) {
             try {
-                this.fetchParams.filterParams = this.widget.definition.fetchParams(this.widget, this.toolbox);
-            } catch (e) {
-                log.error('Error processing fetch params', e);
-                throw new Error('Error processing fetch params', e);
+                this.fetchParams.filterParams = this.widget.definition.fetchParams(
+                    this.widget as Widget<Record<string, keyof Stage.Types.GridParams>>,
```
Comment (Vorbert-Kruk) [inline app/utils/WidgetParamsHandler.ts:75]: This assertions seems to be necessary, looking at the output given by typescript

```bash
TS2345: Argument of type 'StageWidget<GridParams>' is not assignable to parameter of type 'StageWidget<Record<string, unknown>>'.
Type 'GridParams' is not assignable to type 'Record<string, unknown>'.
Index signature is missing in type 'GridParams'.
```

**Review by qooban — COMMENTED** [review]
Looks very good 👍🏻

Added minor comments, I believe they'll be quite easy to fix.

File: app/utils/scriptLoader.ts [hunk 4]
```diff
@@ -1,17 +1,14 @@
-// @ts-nocheck File not migrated fully to TS
 import log from 'loglevel';
 import StageUtils from './stageUtils';
 
 export default class ScriptLoader {
     protected path: string;
 
-    private loaded = false;
-
```
[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply app/utils/scriptLoader.ts:9]: Yeah, precisely
We were only assigning a value to it, without using it 😁

File: app/utils/WidgetBackend.ts [hunk 5]
```diff
@@ -1,12 +1,10 @@
...
 import Internal from './Internal';
 import consts from './consts';
+import type { ManagerData } from '../reducers/managerReducer';
 
 export default class WidgetBackend extends Internal {
-    constructor(widgetId, data) {
+    constructor(private widgetId: string, data: ManagerData) {
         super(data);
-        this.widgetId = widgetId;
     }
```
[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply app/utils/WidgetBackend.ts:8]: Yeah it's a neat shortcut, which I've learned while working with Nestjs ([example](https://github.com/nestjs/nest/blob/master/sample/30-event-emitter/src/orders/orders.controller.ts)) - they use this pattern in every possible place 😄 

But when it comes to our codebase, I've seen that we were already using this pattern - [see example](https://github.com/cloudify-cosmo/cloudify-stage/tree/master/widgets/blueprintSources/src/actions.ts) 💪

File: app/utils/UserAppDataAutoSaver.ts [hunk 6]
```diff
@@ -61,7 +73,7 @@ export default class UserAppDataAutoSaver {
         }
     }
 
-    static create(store) {
+    static create(store: ReduxStore) {
```
[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply app/utils/UserAppDataAutoSaver.ts:76]: When it comes to the timeframe of our Redux meeting and this PR development, one quote is coming to my mind   

image
😄

File: app/utils/StyleLoader.ts [hunk 7]
```diff
@@ -1,9 +1,11 @@
-// @ts-nocheck File not migrated fully to TS
-
 import StageUtils from './stageUtils';
 
 export default class StyleLoader {
-    constructor(stylePath) {
+    path: string;
+
+    loaded: boolean;
```
[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply app/utils/StyleLoader.ts:6]: Oh yeah, nicely caught! 
Removing it right away 🚀

[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply app/utils/StyleLoader.ts]: Removed as a part of 5ad5abdc2dcc51db4f0661d571767a7f2c3b06d4

File: app/utils/Internal.ts [hunk 8]
```diff
@@ -4,7 +4,7 @@ import Consts from './consts';
 import StageUtils from './stageUtils';
 
 export default class Internal extends External {
-    buildHeaders() {
+    buildHeaders(): Record<string, string> {
```
[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply app/utils/Internal.ts]: Seems to me like a good solution 🥳

[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply app/utils/Internal.ts]: Applied as a part of 25fd3127074cfc7dfe84168456b492a3548d9337

File: app/utils/WidgetParamsHandler.ts [hunk 9]
```diff
@@ -1,21 +1,26 @@
...
+    filterParams: QueryStringParams;
+}
 
 export default class WidgetParamsHandler {
-    constructor(widget, toolbox) {
-        this.widget = widget;
-        this.toolbox = toolbox;
+    fetchParams: FetchParams;
 
+    constructor(private widget: Widget<Stage.Types.GridParams>, private toolbox: Toolbox) {
```
[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply app/utils/WidgetParamsHandler.ts]: Woah, I haven't thought about it 🤯 
That's a neat and fully working solution 🎉 
Definitely gonna apply this! 🚀

[the remark this answers is not in the document]
Comment (Vorbert-Kruk) [reply app/utils/WidgetParamsHandler.ts]: Applied as a part of eaaacf1839cfed55d41a0135aeba039461b918b9
Thanks for the awesome suggestion! 🎉

**Review by qooban — APPROVED** [review]
Thanks for all adjustments and clarifications. Well done 👍🏻

**Pull request merged**